src/concrete/ml/sklearn/base.py

```diff
@@ -1688,3 +1692,264 @@ def predict_proba(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) ->
+        # This function can neither be a class method nor a static one because self we want to avoid
+        # having self as a parameter while still being able to access some of its attribute
+        def inference_to_compile(q_X: numpy.ndarray) -> numpy.ndarray:
+            """Compile the circuit in FHE using only the inputs as parameters.
+
+            Args:
+                q_X (numpy.ndarray): The quantized input data
+
+            Returns:
+                numpy.ndarray: The circuit is outputs.
+            """
+            return self._inference(q_X)
+
+        # Create the compiler instance
+        compiler = Compiler(inference_to_compile, {"q_X": "encrypted"})
+
+        return compiler
+
+    @staticmethod
+    def top_k_indices(distance_matrix: numpy.ndarray, k: int) -> numpy.ndarray:
+        """Get the indices of the top-k smallest distances for each point.
+
+        Args:
+            distance_matrix (numpy.ndarray): Represents the pairwise euclidean distance between
+                the query and other points
+            k (int): The top nearest neighbors to consider
+
+        Returns:
+            numpy.ndarray: The k nearest neighbors for the corresponding query, sorted in
+            ascending order.
+        """
+
+        # Sort the distances in an ascending order and select the k smallest distanes
+        return numpy.argsort(distance_matrix, axis=1)[:, :k]
+
+    @staticmethod
+    def majority_vote(nearest_classes: numpy.ndarray):
+        """Determine the most common class among nearest neighborsfor each query.
+
+        Args:
+            nearest_classes (numpy.ndarray): The class labels of the nearest neighbors for a query
+
+        Returns:
+            numpy.ndarray: The majority-voted class label for the corresponding query.
+        """
+        # Get the number of queries (rows) and k (number of nearest points)
+        n_queries, _ = nearest_classes.shape
+        # Compute the majority vote for each query
+        majority_votes = numpy.array([0] * n_queries, dtype=int)
+        for i in range(n_queries):
+            # Use bincount to count occurrences of each class and find the most common one
+            class_counts = numpy.bincount(nearest_classes[i])
+            majority_votes[i] = numpy.argmax(class_counts)
+
+        return majority_votes
+
+    def _inference(self, q_X: numpy.ndarray) -> numpy.ndarray:
+        """Inference function.
+
+        Args:
+            q_X (numpy.ndarray): The quantized input values.
+
+        Returns:
+            numpy.ndarray: The quantized predicted values.
+        """
+        assert self._q_X_fit_quantizer is not None, self._is_not_fitted_error_message()
+
+        # Pairwise euclidean distance
+        # dist(x, y) = sqrt(dot(x, x) - 2 * dot(x, y) + dot(y, y))
+        distance_matrix = (
+            numpy.sum(q_X**2, axis=1, keepdims=True)
+            - 2 * q_X @ self._q_X_fit.T
+            + numpy.expand_dims(numpy.sum(self._q_X_fit**2, axis=1), 0)
```

Comment:
This can be done at training time I suppose

Reply:
it's a constant no? it will be precomputed by CP